### ISSUE TITLE
Control over Muzzleflash and Flashlight functionality

### ISSUE THREAD

**Issue by Servius** [Facepunch/garrysmod-requests#1420]
My focus for GMod is mainly weapons which the above title probably indicates. One thing Id like to be able to control is the flashlight exit position. Id also like more control over the muzzle position, like having more than one muzzle point. Alternating exit locations should be fairly easy as it would just be multiple $attachment points in the compile process. I think having multiple flashlight $attachment points would also be a good step. It was indicated on the GMod Discord that this kind of support would be fairly quick and easy. A low hanging fruit if you will. 

- Multi Muzzle Locations
- Flashlight exit location control
- Multi Flashlight Exit location control

Possible use cases: 
- Weapon Attachments that are Flashlights
- Playermodels having flashlights exit from specific a specific location on it(like a helmet light[s])
- Rotary weapons or multi barrel weapons that could use alternating muzzle positions.

**Comment by Kefta** [comment]
You can already control the muzzleflash yourself on SWEPs using [FireAnimationEvent](http://wiki.garrysmod.com/page/WEAPON/FireAnimationEvent).

**Comment by Servius** [comment]
@Kefta You can have multiple muzzle positions through that?

**Comment by Kefta** [comment]
You can play the muzzle flash effect at whatever attachment you want, even multiple times. That's what the EffectData structure is for.

**Comment by robotboy655** [comment]
You can create completely custom flashlights using ProjectedTextures if you want to, so everything you want is already possible, isn't it?

**Comment by Kefta** [comment]
Using normal projected textures is not the same. Flashlights (can) do a lot of extra clipping and layering, and are controlled by a number of convars that would be extremely inefficient to replicate with Lua rendering hooks.

**Comment by Servius** [comment]
@Kefta Exactly. Having the positioning done as part of the compile process would be extremely simple and wouldnt change much for current content pipelines. Also, you are completely correct regarding the muzzle flash process. I hadnt looked too much into that so thank you for info!

**Comment by robotboy655** [comment]
I don't know how you expect it to work during compile process, the flashlight effects in 3rd person are not attached the the weapon at all as far as I know.

**Comment by Kefta** [comment]
3rd person flashlight position seems to be controlled here: https://github.com/VSES/SourceEngine2007/blob/43a5c90a5ada1e69ca044595383be67f40b33c61/src_main/game/client/c_baseplayer.cpp#L1038

Changing all parts of the first person effect is a bit more difficult since the renderer assumes the light emits from the center of the viewport when determining exposure.
